cross-env, cross-env-shell not working on MacOS.13

Ran across this tool.  Seems to be very useful in its description.  But when I first ran the tool it did not act in a way that I expected.  I expected running the tool in both environments to be exactly the same and they were not.

### Environment
cross-env version: 5.1.0
node version: 8.7.0
npm version: 5.4.2
Windows version: 10 1703 (OS Build 15063.674)
MacOS: 10.13

### Test code
I created 2 simple test scripts and ran them, one for windows the other for MacOS.  Then I ran them 

Windows: `cross-env-test.cmd`:
```
@echo off
echo Argument: %1
echo Environment: %CROSS_ENV_TEST%
```

MacOS: `cross-env-test.sh`
```
#!/bin/bash

echo Argument: $1
echo Environment: $CROSS_ENV_TEST
```

### Results
Windows:
```
E:\>cross-env CROSS_ENV_TEST=Working cross-env-test.cmd $CROSS_ENV_TEST
Argument: "Working"
Environment: Working

E:\>cross-env-shell CROSS_ENV_TEST=Working cross-env-test.cmd $CROSS_ENV_TEST
Argument: Working
Environment: Working
```

MacOS:
```
MacBook:~$ cross-env CROSS_ENV_TEST=Working ./cross-env-test.sh  $CROSS_ENV_TEST
Argument:
Environment: Working
MacBook:~$ cross-env-shell CROSS_ENV_TEST=Working ./cross-env-test.sh  $CROSS_ENV_TEST
Argument:
Environment: Working
```
###  Expected
I expected the results of both operations to be **exactly** the same.

### Suggestion
As a verification I cloned the project on my MacOS and ran all your tests and they passed.  But all of the tests were unit tests you had no full integration tests.
Recommend you add full integration tests, feel free to use the tests I supplied.

@kentcdodds, errr

When I run on the command line it works.  If I put it into a package.json and `npm run` I cannot get it to work.  Note this works fine on Windows.

### Attempt 1

`package.json` script command:
`"foo": "cross-env-shell CROSS_ENV_TEST=Working \"cross-env-test  \$CROSS_ENV_TEST\""`

#### Result
_Cannot handle escaped_ `$`
```
MacBook:cross-env brucejo$ npm run foo
npm ERR! file /Users/brucejo/test/cross-env/package.json
npm ERR! code EJSONPARSE
npm ERR! Failed to parse json
npm ERR! Unexpected token $ in JSON at position 165 while parsing near '... \"cross-env-test  \$CROSS_ENV_TEST\"",
npm ERR! ...'
npm ERR! File: /Users/brucejo/test/cross-env/package.json
npm ERR! Failed to parse package.json data.
npm ERR! package.json must be actual JSON, not just JavaScript.
npm ERR! 
npm ERR! Tell the package author to fix their package.json file. JSON.parse

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/brucejo/.npm/_logs/2017-11-07T18_51_13_676Z-debug.log
```

### Attempt 2: Remove escaping for `$`
`package.json` script command:
`"foo": "cross-env-shell CROSS_ENV_TEST=Working \"cross-env-test  $CROSS_ENV_TEST\""`

#### Result
_Does not substitute the command argument._

```
MacBook:cross-env brucejo$ npm run foo

> MERIS@1.0.0 foo /Users/brucejo/test/cross-env
> cross-env-shell CROSS_ENV_TEST=Working "cross-env-test  $CROSS_ENV_TEST"

Argument:
Environment: Working
```

### Attempt 3: remove enclosing quotes
`package.json` script command:
`"foo": "cross-env-shell CROSS_ENV_TEST=Working cross-env-test  $CROSS_ENV_TEST"`

#### Result
_Does not substitute the command argument._

```
MacBook:cross-env brucejo$ npm run foo

> MERIS@1.0.0 foo /Users/brucejo/test/cross-env
> cross-env-shell CROSS_ENV_TEST=Working "cross-env-test  $CROSS_ENV_TEST"

Argument:
Environment: Working
```

@kentcdodds, I got some more data.

Essentially, what I found was that windows quoting and OSX (bash) quoting were exactly the opposite in terms of what worked.

|Expression | Windows | OSX |
| ----- | ----- | ----- |
| unquoted | works | doesn't work |
| " quoted | works | doesn't work |
| ' quoted | doesn't work | works |

I propose enabling Windows to work with `'` quotes.  See **Proposed Fix**.

### Environment
node - 8.8.1
npm - 5.4.2

### More data

OK, a little more research.  I created these rules in ` package.json`:
```json
    "greet": "npm run greet0 && npm run greet1 && npm run greet2",
    "greet0": "cross-env-shell GREETING=Hi NAME=Joe echo unquoted $GREETING",
    "greet1": "cross-env-shell GREETING=Hi NAME=Joe \"echo doubleQuotes $GREETING\"",
    "greet2": "cross-env-shell GREETING=Hi NAME=Joe 'echo singleQuotes $GREETING'"
```


## On OSX
For OSX I also added a console.log to output the return value from commandConvert in [command.js](https://github.com/kentcdodds/cross-env/blob/master/src/command.js#L16).
```
MacBook:cross-env brucejo$ npm run greet

> MERIS@1.0.0 greet /Users/brucejo/test/cross-env
> npm run greet0 && npm run greet1 && npm run greet2


> MERIS@1.0.0 greet0 /Users/brucejo/test/cross-env
> cross-env-shell GREETING=Hi NAME=Joe echo unquoted $GREETING

commandConvert: echo
commandConvert: unquoted
unquoted

> MERIS@1.0.0 greet1 /Users/brucejo/test/cross-env
> cross-env-shell GREETING=Hi NAME=Joe "echo doubleQuotes $GREETING"

commandConvert: echo doubleQuotes 
doubleQuotes

> MERIS@1.0.0 greet2 /Users/brucejo/test/cross-env
> cross-env-shell GREETING=Hi NAME=Joe 'echo singleQuotes $GREETING'

commandConvert: echo singleQuotes $GREETING
singleQuotes Hi
```

The only one that properly processes the `$GREETING` environment variable is the singleQuotes.  The others look like they evaluate at it in the calling shell.

## On Windows
For Windows I also added a console.log to output the return value from commandConvert in [command.js](https://github.com/kentcdodds/cross-env/blob/master/src/command.js#L27).
```
E:\temp\cross-env>npm run greet

> foo@1.0.0 greet E:\temp\cross-env
> npm run greet0 && npm run greet1 && npm run greet2


> foo@1.0.0 greet0 E:\temp\cross-env
> cross-env-shell GREETING=Hi NAME=Joe echo unquoted $GREETING

commandConvert: echo
commandConvert: unquoted
commandConvert: %GREETING%
unquoted Hi

> foo@1.0.0 greet1 E:\temp\cross-env
> cross-env-shell GREETING=Hi NAME=Joe "echo doubleQuotes $GREETING"

commandConvert: echo doubleQuotes %GREETING%
doubleQuotes Hi

> foo@1.0.0 greet2 E:\temp\cross-env
> cross-env-shell GREETING=Hi NAME=Joe 'echo singleQuotes $GREETING'

commandConvert: 'echo
commandConvert: singleQuotes
commandConvert: %GREETING%'
''echo' is not recognized as an internal or external command,
operable program or batch file.
events.js:182
      throw er; // Unhandled 'error' event
      ^

Error: spawn 'echo ENOENT
    at notFoundError (E:\temp\cross-env\node_modules\cross-spawn\lib\enoent.js:11:11)
    at verifyENOENT (E:\temp\cross-env\node_modules\cross-spawn\lib\enoent.js:46:16)
    at ChildProcess.cp.emit (E:\temp\cross-env\node_modules\cross-spawn\lib\enoent.js:33:19)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! foo@1.0.0 greet2: `cross-env-shell GREETING=Hi NAME=Joe 'echo singleQuotes $GREETING'`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the foo@1.0.0 greet2 script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Bruce\AppData\Roaming\npm-cache\_logs\2017-11-09T05_44_12_699Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! foo@1.0.0 greet: `npm run greet0 && npm run greet1 && npm run greet2`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the foo@1.0.0 greet script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Bruce\AppData\Roaming\npm-cache\_logs\2017-11-09T05_44_12_720Z-debug.log
```
Essentially, Windows works with all forms but using the single quotes fails it.

## Proposed Fix
If I look at the Windows output, essentially Windows treats `'` like any other character.  And what commandConvert finds is `commandConvert: 'echo`.

Then it tries to spawn `'echo`.

I propose adding code to look at each argument to [`parseCommand`](https://github.com/kentcdodds/cross-env/blob/master/src/index.js#L35) and remove any leading or trailing `'` character.  Here is the code:

```
function parseCommand(_args) {
  // Windows does not support nested quotes via ''
  // *nix does.  So ignore any found ' at the
  // beginning or end of an argument.
  var args = _args.map(function (_a) {
    var a = _a;
    if(a[0] === "'") a = a.slice(1, a.length);
    if(a[a.length - 1] === "'") a = a.slice(0, a.length - 1);
    return a;
  });
```

@kentcdodds, let me know if this looks good to you and I can gin up a PR.

--- a/src/__tests__/index.js
+++ b/src/__tests__/index.js
@@ -74,6 +74,23 @@ it(`should handle quoted scripts`, () => {
   })
 })
 
+it(`should handle escaped characters`, () => {
+  // this escapes \,",' and $
+  crossEnv(['GREETING=Hi', 'NAME=Joe', 'echo \\"\\\'\\$GREETING\\\'\\" && echo $NAME'], {
+    shell: true,
+  })
+  expect(
+    crossSpawnMock.spawn,
+  ).toHaveBeenCalledWith("echo \"'$GREETING'\" && echo $NAME", [], {
+    stdio: 'inherit',
+    shell: true,
+    env: Object.assign({}, process.env, {
+      GREETING: 'Hi',
+      NAME: 'Joe',
+    }),
+  })
+})
+
 it(`should do nothing given no command`, () => {
   crossEnv([])
   expect(crossSpawnMock.spawn).toHaveBeenCalledTimes(0)

--- a/src/index.js
+++ b/src/index.js
@@ -51,8 +51,21 @@ function parseCommand(args) {
       envSetters[match[1]] = value
     } else {
       // No more env setters, the rest of the line must be the command and args
-      command = args[i]
-      commandArgs = args.slice(i + 1)
+      let cStart = []
+      cStart = args.slice(i)
+        // Regex:
+        // match "\'" or "'"
+        // or match "\" if followed by [$"\] (lookahead)
+        .map((a) => {
+        const re = new RegExp(/(\\)?'|([\\])(?=[$"\\])/, 'g')
+        // Eliminate all matches except for "\'" => "'"
+        return a.replace(re, (m) => {
+          if(m === "\\'") return "'"
+          return ""
+        })
+      })
+      command = cStart[0]
+      commandArgs = cStart.slice(1)
       break
     }
   }

